exporters/zipkin/src/test/java/io/opentelemetry/exporters/zipkin/ZipkinSpanExporterEndToEndHttpTest.java

```diff
@@ -0,0 +1,195 @@
+
+    exportAndVerify(configuration);
+  }
+
+  @Test
+  public void testExportAsThrift() {
+
+    @SuppressWarnings("deprecation")
+    ZipkinExporterConfiguration configuration =
+        ZipkinExporterConfiguration.builder()
+            .setSender(
+                URLConnectionSender.newBuilder()
+                    .endpoint(zipkin.httpUrl() + ENDPOINT_V1_SPANS)
+                    .encoding(Encoding.THRIFT)
+                    .build())
+            .setServiceName(SERVICE_NAME)
+            .setEncoder(SpanBytesEncoder.THRIFT)
+            .build();
+
+    exportAndVerify(configuration);
+  }
+
+  @Test
+  public void testExportAsJsonV1() {
+
+    ZipkinExporterConfiguration configuration =
+        ZipkinExporterConfiguration.builder()
+            .setSender(
+                URLConnectionSender.newBuilder()
+                    .endpoint(zipkin.httpUrl() + ENDPOINT_V1_SPANS)
+                    .encoding(Encoding.JSON)
+                    .build())
+            .setServiceName(SERVICE_NAME)
+            .setEncoder(SpanBytesEncoder.JSON_V1)
+            .build();
+
+    exportAndVerify(configuration);
+  }
+
+  @Test
+  public void testExportFailedAsWrongEncoderUsed() {
+
+    ZipkinExporterConfiguration configuration =
+        ZipkinExporterConfiguration.builder()
+            .setSender(
+                URLConnectionSender.newBuilder()
+                    .endpoint(zipkin.httpUrl() + ENDPOINT_V2_SPANS)
+                    .encoding(Encoding.JSON)
+                    .build())
+            .setServiceName(SERVICE_NAME)
+            .setEncoder(SpanBytesEncoder.PROTO3)
+            .build();
+
+    ZipkinSpanExporter zipkinSpanExporter = ZipkinSpanExporter.create(configuration);
+
+    SpanData spanData = buildStandardSpan().build();
+    SpanExporter.ResultCode resultCode = zipkinSpanExporter.export(Collections.singleton(spanData));
+
+    assertThat(resultCode).isEqualTo(SpanExporter.ResultCode.FAILURE);
+    List<zipkin2.Span> zipkinSpans = zipkin.getTrace(TRACE_ID);
+    assertThat(zipkinSpans).isNotNull();
+    assertThat(zipkinSpans).isEmpty();
+  }
+
+  /**
+   * Exports a span, verify that it was received by Zipkin, and check that the span stored by Zipkin
+   * matches what was sent.
+   */
+  private void exportAndVerify(ZipkinExporterConfiguration configuration) {
+    ZipkinSpanExporter zipkinSpanExporter = ZipkinSpanExporter.create(configuration);
+
+    SpanData spanData = buildStandardSpan().build();
+    SpanExporter.ResultCode resultCode = zipkinSpanExporter.export(Collections.singleton(spanData));
+
+    assertThat(resultCode).isEqualTo(SpanExporter.ResultCode.SUCCESS);
+    List<zipkin2.Span> zipkinSpans = zipkin.getTrace(TRACE_ID);
+
+    assertThat(zipkinSpans).isNotNull();
+    assertThat(zipkinSpans.size()).isEqualTo(1);
+    assertThat(zipkinSpans.get(0))
+        .isEqualTo(ZipkinSpanExporter.generateSpan(spanData, localEndpoint));
+  }
+
+  private static SpanData.Builder buildStandardSpan() {
```

Comment:
typically when I write tests like this, I add a second method directly over and under it so that you can see the translation. This helps people understand the intent and any surprising things such as annotation naming choices or missed timestamp resolutions.

to do that, I would move attributes and annotations to locals here so that you can see everything without too much scrolling. the ID constants which are fine to be up top and plus you'd use these for zipkin also.

Reply:
I've not gone overboard on the tidying up, but I added a method to build the Zipkin span underneath the OTel span